src/geometry/euclidean/congruence.lean

```diff
@@ -0,0 +1,303 @@
+    rw [← set.apply_range_splitting v₁ fx, ← set.apply_range_splitting v₁ fy],
+    rw h.to_dist_eq_dist,
+    refl,
+  end}
+
+lemma isometry_sound (h : v₁ ≅ v₂) (i : ι) :
+  ↑(h.isometry (set.range_factorization v₁ i)) = v₂ i :=
+h.map_sound i
+
+
+
+
+
+variables {V₁ V₂ : Type*} [inner_product_space ℝ  V₁] [inner_product_space ℝ V₂]
+  [normed_add_torsor V₁ P₁] [normed_add_torsor V₂ P₂]
+
+
+
+
+
+include V₁ V₂
+
+lemma to_angle_eq_angle (h : v₁ ≅ v₂) (i₁ i₂ i₃ : ι) :
+  ∠ (v₁ i₁) (v₁ i₂) (v₁ i₃) = ∠ (v₂ i₁) (v₂ i₂) (v₂ i₃) :=
+begin
+  unfold euclidean_geometry.angle, unfold inner_product_geometry.angle,
+  have key := abs_le.1 (abs_real_inner_div_norm_mul_norm_le_one _ _),
+  have key':= abs_le.1 (abs_real_inner_div_norm_mul_norm_le_one _ _),
+  rw real.arccos_inj key.1 key.2 key'.1 key'.2,
+  simp [real_inner_eq_norm_mul_self_add_norm_mul_self_sub_norm_sub_mul_self_div_two,
+    ← normed_add_torsor.dist_eq_norm', h.to_dist_eq_dist],
+end
+
+omit V₁ V₂
+
+variables {a b c : P₁} {d e f : P₂}
+
+
+lemma _root_.fin.fin_3_swap_left {α : Type*} (a b c : α) : ![a,b,c] = ![b,a,c] ∘ ![1,0,2] :=
+by {ext x, fin_cases x; refl}
+lemma _root_.fin.fin_3_swap_right {α : Type*} (a b c : α) : ![a,b,c] = ![a,c,b] ∘ ![0,2,1] :=
+by {ext x, fin_cases x; refl}
+
+
+lemma triangle_swap_left (h : ![a,b,c] ≅ ![d,e,f]) : ![b,a,c] ≅ ![e,d,f] :=
+begin
+  rw [fin.fin_3_swap_left b a c, fin.fin_3_swap_left e d f],
+  apply sub_congruence ![1,0,2] h,
+end
+lemma triangle_swap_right (h : ![a,b,c] ≅ ![d,e,f]) : ![a,c,b] ≅ ![d,f,e] :=
+begin
+  rw [fin.fin_3_swap_right a c b, fin.fin_3_swap_right d f e],
+  apply sub_congruence ![0,2,1] h,
+end
+
+lemma side_side_side {a b c : P₁} {d e f : P₂} (hs₁ : dist a b = dist d e)
+  (hs₂ : dist a c = dist d f) (hs₃ : dist b c = dist e f) : ![a,b,c] ≅ ![d,e,f] :=
+begin
+  intro i₁, fin_cases i₁; intro i₂; fin_cases i₂;
+    simp [dist_comm]; assumption,
+end
+
+include V₁
+
+lemma cos_angle_elim : real.cos (∠ a b c) =
+  (dist a b * dist a b + dist b c * dist b c - dist a c * dist a c) / (2 * dist a b * dist b c) :=
+begin
+  unfold euclidean_geometry.angle, rw inner_product_geometry.cos_angle,
+  rw real_inner_eq_norm_mul_self_add_norm_mul_self_sub_norm_sub_mul_self_div_two,
+  simp [ ← normed_add_torsor.dist_eq_norm'],
+  rw [div_div, mul_assoc, dist_comm b c],
+end
+
+
+
+lemma sin_angle_elim : real.sin (∠ a b c) =
+  real.sqrt (1 - ((dist a b ^ 2 + dist b c ^ 2 - dist a c ^ 2) ^ 2) /
+    (2 * dist a b * dist b c) ^ 2) :=
+begin
+  rw real.sin_eq_sqrt_one_sub_cos_sq
+    (euclidean_geometry.angle_nonneg a b c)
+    (euclidean_geometry.angle_le_pi a b c),
+  rw [cos_angle_elim, ← pow_two, ← pow_two, ← pow_two, div_pow],
+end
+
+omit V₁
+
+/-
+the library contains a similar lemma, but it requires the numerator to be positive instead of
+the denominator, so this is more useful in this case.
+-/
+lemma _root_.real.sqrt_div_sq :
+  ∀ {y : ℝ}, 0 ≤ y → ∀ (x : ℝ), real.sqrt (x / y^2) = real.sqrt x / y :=
```

Comment:
This certainly doesn't belong in a congruence file; if it belongs in mathlib, it should be under `data.real`.

Reply:
Thanks for all the feedback. I've added the appropriate lemma to `data.real.sqrt` in a separate PR.

Reply:
Now that some things here depend on a lemma included in a new PR, how do I import the new version of mathlib into this branch? Or is it easier to just create a new branch with a new PR?

Reply:
`git merge origin/master; git push origin congruences` (adjust as needed if you're using different naming locally).

Reply:
(Make sure to do `git pull` first so that your `origin/master` has the latest upstream changes.)